:art: Refactored code to improve maintainability

andreihava-okta opened the pull request:
Refactored most code into separate classes to improve project maintainability.

smashling commented:
This looks beautiful! I am going to test it now to make sure all the functionality works.

smashling commented:
There are a lot of changes here, so this code review is going to take a little while.

andreihava-okta commented:
@smashling Yeah, I started with "let's move this method" and then I couldn't stop. Thanks for looking at it!

smashling commented:
I dont think .okta should be in the .aws folder, as that creates a new profile store for users that are already using the tool, forcing them to move and breaking current functionality.

AlainODea commented on src/main/java/com/okta/tools/OktaAwsCliEnvironment.java:
```diff
@@ -0,0 +1,12 @@
+    public static String oktaAwsAppUrl;
+
+    public static String awsRoleToAssume;
+}
```
:art: These are mutable global variables.

It would be preferable for each of the individual classes to take a **OktaAwsCliEnvironment** as a constructor argument and have it as a field and then uses of each of the classes are constructed and interacted with as instances.

AlainODea commented on src/main/java/com/okta/tools/authentication/OktaAuthentication.java:
```diff
@@ -0,0 +1,136 @@
+import java.nio.charset.StandardCharsets;
+import java.util.Scanner;
+
+public final class OktaAuthentication {
```
👍 I like this. Moving authentication into it's own class is really good.

smashling replied:
The tool used to reprompt if a bad username/password is inputted, but now gives this error instead, and crashes.
Exception in thread "main" java.lang.RuntimeException: Failed : HTTP error code : 401
	at com.okta.tools.authentication.OktaAuthentication.primaryAuthFailureHandler(OktaAuthentication.java:106)
	at com.okta.tools.authentication.OktaAuthentication.getPrimaryAuthResponse(OktaAuthentication.java:51)
	at com.okta.tools.authentication.OktaAuthentication.getOktaSessionToken(OktaAuthentication.java:30)
	at com.okta.tools.saml.OktaSaml.getSamlResponse(OktaSaml.java:33)
	at com.okta.tools.OktaAwsCliAssumeRole.run(OktaAwsCliAssumeRole.java:62)
	at com.okta.tools.WithOkta.main(WithOkta.java:23)
15:03:33 [main] ERROR com.okta.tools.authentication.OktaAuthentication - Invalid username or password.

AlainODea commented on src/main/java/com/okta/tools/authentication/OktaMFA.java:
```diff
@@ -0,0 +1,433 @@
+import java.util.List;
+import java.util.Scanner;
+
+public class OktaMFA {
```
👍 I like this. Moving MFA into it's own class is really good.

AlainODea commented on src/main/java/com/okta/tools/helpers/ConfigHelper.java:
```diff
@@ -0,0 +1,56 @@
+     * @return The file reader for the config file
+     * @throws FileNotFoundException
+     * @author Andrei Hava
+     * @since 02/14/2018
```
:art: attribution is unnecessary. It's in the Git history.

AlainODea commented on src/main/java/com/okta/tools/helpers/ConfigHelper.java:
```diff
@@ -0,0 +1,56 @@
+     * @return The FileWriter for the config file
+     * @throws IOException
+     * @author Andrei Hava
+     * @since 02/14/2018
```
:art: attribution is unnecessary. It's in the Git history.

AlainODea commented on src/main/java/com/okta/tools/helpers/CredentialsHelper.java:
```diff
@@ -0,0 +1,60 @@
+     * @return The file reader for the credentials file
+     * @throws FileNotFoundException
+     * @author Andrei Hava
+     * @since 02/14/2018
```
:art: attribution is unnecessary. It's in the Git history.

AlainODea commented on src/main/java/com/okta/tools/helpers/CredentialsHelper.java:
```diff
@@ -0,0 +1,60 @@
+     * @return The FileWriter for the credentials file
+     * @throws IOException
+     * @author Andrei Hava
+     * @since 02/14/2018
```
:art: attribution is unnecessary. It's in the Git history.

AlainODea commented on src/main/java/com/okta/tools/helpers/FileHelper.java:
```diff
@@ -0,0 +1,89 @@
+     * @param path The path of the directory to be returned
+     * @return The path of the directory
+     * @author Andrei Hava
+     * @since 02/14/2018
```
:art: attribution is unnecessary. It's in the Git history.

AlainODea commented on src/main/java/com/okta/tools/helpers/FileHelper.java:
```diff
@@ -0,0 +1,89 @@
+     *
+     * @return The path of the AWS directory
+     * @author Andrei Hava
+     * @since 02/14/2018
```
:art: attribution is unnecessary. It's in the Git history.

AlainODea commented on src/main/java/com/okta/tools/helpers/FileHelper.java:
```diff
@@ -0,0 +1,89 @@
+     *
+     * @return The path of the Okta directory
+     * @author Andrei Hava
+     * @since 03/02/2018
```
:art: attribution is unnecessary. It's in the Git history.

AlainODea commented on src/main/java/com/okta/tools/helpers/FileHelper.java:
```diff
@@ -0,0 +1,89 @@
+     * @return The reader for the given file
+     * @throws FileNotFoundException
+     * @author Andrei Hava
+     * @since 02/14/2018
```
:art: attribution is unnecessary. It's in the Git history.

AlainODea commented on src/main/java/com/okta/tools/helpers/FileHelper.java:
```diff
@@ -0,0 +1,89 @@
+     * @return The FileReader for the given path
+     * @throws IOException
+     * @author Andrei Hava
+     * @since 02/14/2018
```
:art: attribution is unnecessary. It's in the Git history.

AlainODea commented on src/main/java/com/okta/tools/helpers/FileHelper.java:
```diff
@@ -0,0 +1,89 @@
+     * @since 02/14/2018
+     */
+    public static Path getAwsDirectory() {
+        return getDirectory(USER_HOME + "/.aws/");
```
:bug: It's unsafe to build paths with String concatenation.

Recommendation: move the pre-existing Path-based methods from OktaAwsCliAssumeRole.

AlainODea commented on src/main/java/com/okta/tools/helpers/FileHelper.java:
```diff
@@ -0,0 +1,89 @@
+     * @since 03/02/2018
+     */
+    public static Path getOktaDirectory() {
+        return getDirectory(getAwsDirectory().toString() + "/.okta/");
```
:bug: It's unsafe to build paths with String concatenation.

Recommendation: move the pre-existing Path-based methods from OktaAwsCliAssumeRole.

AlainODea replied:
:bug: breaking change. The .okta folder needs to remain in ~/.okta or it will break existing users. This is a non-starter.

AlainODea commented on src/main/java/com/okta/tools/helpers/SessionHelper.java:
```diff
@@ -0,0 +1,110 @@
+     *
+     * @return The Path of the current session file
+     * @author Andrei Hava
+     * @since 02/14/2018
```
:art: attribution is unnecessary. It's in the Git history.

rdegges commented:
@mraible wanna take a look at this since you're the Java dude?

AlainODea commented on src/main/java/com/okta/tools/OktaAwsCliAssumeRole.java:
```diff
@@ -15,937 +15,67 @@
 final class OktaAwsCliAssumeRole {
 
     private static final Logger logger = LogManager.getLogger(OktaAwsCliAssumeRole.class);
-    private static final String OKTA_AWS_CLI_SESSION_FILE = ".okta-aws-cli-session";
```
👍 While this is technically a breaking change, it's in the context of a single generated session and I prefer your change of moving this into the **.okta** folder.

AlainODea commented on src/main/java/com/okta/tools/helpers/ConfigHelper.java:
```diff
@@ -0,0 +1,56 @@
+     * @since 02/14/2018
+     */
+    public static Reader getConfigReader() throws FileNotFoundException {
+        return FileHelper.getReader(FileHelper.getAwsDirectory().toString() + "/config");
```
:bug: It's unsafe to build paths with String concatenation.

Recommendation: move the pre-existing Path-based methods from OktaAwsCliAssumeRole.

AlainODea replied:
All of the file path concatenation issues can probably be solved using the strategy I followed here:
https://github.com/oktadeveloper/okta-aws-cli-assume-role/blob/master/src/main/java/com/okta/tools/OktaAwsCliAssumeRole.java#L558-L586

It's a bit more verbose, but I think the explicitness is good and it's hard to work around the conditionally creating folders piece otherwise.

AlainODea commented on src/main/java/com/okta/tools/helpers/ConfigHelper.java:
```diff
@@ -0,0 +1,56 @@
+     * @since 02/14/2018
+     */
+    public static FileWriter getConfigWriter() throws IOException {
+        return FileHelper.getWriter(FileHelper.getAwsDirectory().toString() + "/config");
```
:bug: It's unsafe to build paths with String concatenation.

Recommendation: move the pre-existing Path-based methods from OktaAwsCliAssumeRole.

AlainODea commented:
@andreihava-okta I overall really like the direction this change has taken the code in. The increased modularity will make it much easier to maintain.

rdegges commented:
@AlainODea if you're cool with it I say let's merge! =D

AlainODea commented:
@rdegges the moving of the **.okta** folder into **.aws** will break things for existing users.

The authentication re-attempt crash @smashling found is also fatal.

Can those be fixed prior to merge? We'll have a broken release otherwise.

andreihava-okta commented:
@AlainODea I’ll look into this early tomorrow and modify the code according to your comments. 

Thanks for looking into this and for the great feedback!

AlainODea commented:
@andreihava-okta thank you! I'm looking forward to it.

andreihava-okta commented:
@AlainODea @smashling I've adapted the code to the reviews you've left on it.

Most notable changes:
- Moved .okta back to user home
- Fixed re-authentication crash
- Fixed multiple account logout issue caused by a FileReader not being closed
- Made OktaAwsCliEnvironment a constructor param for most classes, making them no longer static

AlainODea commented on src/main/java/com/okta/tools/OktaAwsCliAssumeRole.java:
```diff
@@ -36,46 +36,71 @@
+        currentSession = sessionHelper.getCurrentSession();
+
+        if (currentSession.isPresent()) {
+            environment.oktaProfile = currentSession.get().profileName;
```
After this point, **environment.oktaProfile** is not set to the correct value for the profile specified by **OKTA_PROFILE**. As a result, multiple profile selection doesn't work.

Right now, the named profile support doesn't work if there is also a session file. The expected behavior is if **OKTA_PROFILE** is not blank, then multiple-profile support should be used to pick the session. Only if it is blank should the legacy session file be used. The way the old system achieved this was to overwrite the legacy session file after determining the profile name from named multi-profile support.

There really should be more unit tests to cover these cases. Sorry about that.

Please try some tests with OKTA_PROFILE set to different values. You should be able to switch between the profiles without re-authenticating, and create new profiles only by seeing the role selection menu.

```bash
env OKTA_PROFILE=dev withokta sts get-caller-identity
env OKTA_PROFILE=prod awscli sts get-caller-identity
env OKTA_PROFILE=dev awscli sts get-caller-identity
env OKTA_PROFILE=prod awscli sts get-caller-identity
```

Make sure you have two different roles to assume. Expect the Okta authentication only once, and the role selection to appear twice times. Check that the get-caller-identity returns the expected assumed-role ARN.

Example:
```plain
arn:aws:sts::123456789012:assumed-role/Admin/botocore-session-1234567890
```

andreihava-okta commented:
@AlainODea The multiple profiles not being used bug has been fixed in the latest commit! 😺

AlainODea commented:
@mraible this is ready to merge.
